chainstate/tx-verifier/src/transaction_verifier/signature_destination_getter.rs

```diff
@@ -78,150 +82,157 @@ impl<'a> SignatureDestinationGetter<'a> {
-                        | AccountCommand::LockTokenSupply(token_id)
-                        | AccountCommand::FreezeToken(token_id, _)
-                        | AccountCommand::UnfreezeToken(token_id)
-                        | AccountCommand::ChangeTokenAuthority(token_id, _) => {
-                            let token_data = tokens_view.get_token_data(token_id)?.ok_or(
-                                SignatureDestinationGetterError::TokenDataNotFound(*token_id),
-                            )?;
-                            let destination = match token_data {
-                                tokens_accounting::TokenData::FungibleToken(data) => {
-                                    data.authority().clone()
-                                }
-                            };
-                            Ok(destination)
+                        TxOutput::CreateStakePool(_, pool_data) => {
+                            // Spending an output of a pool creation transaction is only allowed in a
+                            // context of a transaction (as opposed to block reward) only if this pool
+                            // is being decommissioned.
+                            // If this rule is being invalidated, it will be detected in other parts
+                            // of the code.
+                            Ok(pool_data.decommission_key().clone())
+                        }
+                        TxOutput::ProduceBlockFromStake(_, pool_id) => {
+                            // The only way we can spend this output in a transaction
+                            // (as opposed to block reward), is if we're decommissioning a pool.
+                            Ok(accounting_view
+                                .get_pool_data(*pool_id)?
+                                .ok_or(SignatureDestinationGetterError::PoolDataNotFound(*pool_id))?
+                                .decommission_destination()
+                                .clone())
+                        }
+                        TxOutput::IssueNft(_, _, d) => Ok(d.clone()),
+                        TxOutput::IssueFungibleToken(_) => {
+                            // This error is emitted in other places for attempting to make this spend,
+                            // but this is just a double-check.
+                            Err(SignatureDestinationGetterError::SigVerifyOfNotSpendableOutput)
                         }
-                    },
+                        TxOutput::AtomicTransfer(_, htlc) => match input_witness {
+                            InputWitness::NoSignature(_) => todo!(),
+                            InputWitness::Standard(sig) => {
+                                let destination = AuthorizedPublicKeyHashSecretSpend::from_data(
+                                    sig.raw_signature(),
+                                )
+                                .map_or(htlc.refund_key, |_| htlc.spend_key);
```

Comment:
There probably should be some enum inside that struct that tells you which arm was chosen.